modules/keyword_generator.py

```diff
@@ -0,0 +1,158 @@
+        cves (Optional[List[str]]): List of CVEs.
+    
+    Returns:
+        List[str]: List of keywords generated from the source.
+    """
+    all_keywords = set()
+    seen_products = set()
+    
+    if isinstance(source, str) and version:
+        # Source is a product with a version
+        all_keywords.update(generate_keyword_list_from_product(source, version, seen_products))
+    elif isinstance(source, list) and all(isinstance(item, str) for item in source):
+        # Source is a list of CVEs
+        all_keywords.update(generate_keywords_from_cves(source))
+    elif isinstance(source, list) and all(isinstance(item, (list, tuple)) for item in source):
+        # Source is a host array
+        all_keywords.update(generate_keywords_list_from_host_array(source, cves))
+    else:
+        raise ValueError("Invalid source type")
+    
+    return list(all_keywords)
+
+def generate_keywords_with_ai(api_key: str, source: Union[str, List[Union[List, tuple]], List[str]], version: Optional[str] = None, cves: Optional[List[str]] = None) -> List[str]:
+    """
+    Generate keywords from different sources by sending all the data to the OpenAI model.
+    
+    Args:
+        api_key (str): The OpenAI API key.
+        source (Union[str, List[Union[List, tuple]], List[str]]): The source of keywords.
+        version (Optional[str]): The version name if source is a product.
+        cves (Optional[List[str]]): List of CVEs.
+    
+    Returns:
+        List[str]: List of generated keywords.
+    """
+    openai.api_key = api_key
+    
+    # Create a detailed prompt based on the input data
+    prompt_parts = ["give me a list of CVE's based on the following data:"]
+    
+    if isinstance(source, str) and version:
+        prompt_parts.append(f"Product: {source}, Version: {version}")
+    
+    if isinstance(source, list) and all(isinstance(item, str) for item in source):
+        prompt_parts.append(f"CVEs: {', '.join(source)}")
+    
+    if isinstance(source, list) and all(isinstance(item, (list, tuple)) for item in source):
+        for port_info in source:
+            if len(port_info) >= 5:
+                port, proto, product, service, version = port_info[:5]
+                prompt_parts.append(f"Port: {port}, Protocol: {proto}, Product: {product}, Service: {service}, Version: {version}")
+    
+    if cves:
+        prompt_parts.append(f"Additional CVEs: {', '.join(cves)}")
+    
+    prompt = "\n".join(prompt_parts)
+    
+    prompt += "\nONLY RESPONSES IN THE FORMAT 'CVE-XXXX-XXXX, CVE-XXXX-XXXX, ...' WILL BE ACCEPTED"
+    
+    # Send the prompt to the OpenAI API
+    response = openai.chat.completions.create(
+        model="gpt-4o",
+        messages=[
+            {"role": "user", "content": prompt}
+        ],
+    )
+    
+    # Extract and return the keywords from the response
+    keywords = response.choices[0].message.content.strip().split(', ')
+    return keywords
```

Comment:
I don't have an openai api key. Have you tested this? If you did can you inform me about how it works.

Reply:
yeah, I can confirm it works, the prompt is the main code of this, I'd say 99.9% of the time it sticks to it and works great, 0.1 itll say something before giving the CVEs and itll try and run "hello" as a keyword, no harm but yk

Reply:
Alright very cool work